db/set.go

```diff
@@ -116,3 +220,158 @@ func (set *Set) SMembers() ([][]byte, error) {
+	return set.meta.Len, nil
+}
+
+// SIsmember returns if member is a member of the set stored at key
+func (set *Set) SIsmember(member []byte) (int64, error) {
+	if !set.Exists() {
+		return 0, nil
+	}
+	dkey := DataKey(set.txn.db, set.meta.ID)
+	ikey := setItemKey(dkey, member)
+
+	value, err := set.txn.t.Get(ikey)
+	if err != nil {
+		if IsErrNotFound(err) {
+			return 0, nil
+		}
+		return 0, err
+	}
+	if !bytes.Equal(value, SetNilValue) {
+		return 0, nil
+	}
+	return 1, nil
+}
+
+// SPop removes and returns one or more random elements from the set value store at key.
+func (set *Set) SPop(count int64) (members [][]byte, err error) {
+	if !set.Exists() || set.meta.Len == 0 {
+		return nil, nil
+	}
+	dkey := DataKey(set.txn.db, set.meta.ID)
+	prefix := append(dkey, ':')
+	iter, err := set.txn.t.Iter(prefix, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Close()
+	var del int64
+	var ms [][]byte
+	for iter.Valid() && iter.Key().HasPrefix(prefix) {
+		if count == 0 {
+			ms = append(ms, iter.Key()[len(prefix):])
+			if err := set.txn.t.Delete([]byte(iter.Key())); err != nil {
+				return nil, err
+			}
+			set.meta.Len--
+			break
+		} else {
+			ms = append(ms, iter.Key()[len(prefix):])
+			if err := set.txn.t.Delete([]byte(iter.Key())); err != nil {
+				return nil, err
+			}
+			del++
+			count--
+			if count == 0 {
+				set.meta.Len -= del
+				break
+			}
+			if err := iter.Next(); err != nil {
+				return nil, err
+			}
+		}
+	}
+	if count != 0 {
+		set.meta.Len = 0
+	}
+	if err := set.updateMeta(); err != nil {
+		return nil, err
+	}
+	return ms, nil
+}
+
+// SRem removes the specified members from the set stored at key
+func (set *Set) SRem(members [][]byte) (int64, error) {
+	var num int64
+	if !set.Exists() {
+		return 0, nil
+	}
+	dkey := DataKey(set.txn.db, set.meta.ID)
+	ms := RemoveRepByMap(members)
+	ikeys := make([][]byte, len(ms))
+	for i := range ms {
+		ikeys[i] = setItemKey(dkey, ms[i])
+		value, err := set.txn.t.Get(ikeys[i])
+		if err != nil {
+			if IsErrNotFound(err) {
+				continue
+			}
+			return 0, err
+		}
+		if bytes.Equal(value, SetNilValue) {
+			if err := set.txn.t.Delete([]byte(ikeys[i])); err != nil {
+				return 0, err
+			}
+			num++
+		}
+	}
+	set.meta.Len -= num
+	if err := set.updateMeta(); err != nil {
+		return 0, err
+	}
+	return num, nil
+}
+
+// SMove movies member from the set at source to the set at destination
+func (set *Set) SMove(destination []byte, member []byte) (int64, error) {
+
+	if !set.Exists() {
+		return 0, nil
+	}
+	res, err := set.SIsmember(member)
+	if err != nil {
+		return 0, err
+	}
+	if res == 0 {
+		return 0, nil
+	}
+	destset, _ := GetSet(set.txn, destination)
+	res, err = destset.SIsmember(member)
+	if err != nil {
+		return 0, err
+	}
+	if res == 0 {
+		if _, err := destset.SAdd(member); err != nil {
+			return 0, err
+		}
+		destset.meta.Len++
+	}
+	dkey := DataKey(set.txn.db, set.meta.ID)
+	ikey := setItemKey(dkey, member)
+
+	value, err := set.txn.t.Get(ikey)
```

Comment:
These lines issued redundant requests to TiKV which has been issued by SIsmember.

Reply:
Removed some code that overlaps with Ismembers